src/index.ts

```diff
@@ -162,6 +165,28 @@ async function main() {
   } catch (exception) {
     logger.error({ exception }, 'BlockchainReader was unable to start')
   }
+
+  return {
+    stop: async () => {
+      logger.info('Po.et node terminating...')
+
+      await api.stop()
+      await blockchainReader.stop()
+      await batchReader.stop()
+      await view.stop()
+      await batchWriter.stop()
+      await blockchainWriter.stop()
+      await storageWriter.stop()
+      await storage.stop()
+
+      logger.info('Po.et node terminated')
+      return true
+    },
+  }
 }
 
-main().catch(console.error)
+/* istanbul ignore if */
+if (!module.parent)
```

Comment:
I don't think we should have logic for coverage code in the production code. Can we get rid of this by extracting the code that starts the app from this file? I think we should do that anyway. The file that composes the server together should not also be starting the server. That allows us to import the server to any other file and start it in many different ways for different things. I would be ok if we tackle that in a different PR that as I am not sure how much change it would require for the coverage runner.

Reply:
Sorry, I'm not following. This is the entry point for the package. Where would a call to main.start() be extracted to?

This is not primarily code for instrumenting coverage. 

When `npm start` is run, main.start() gets invoked since the module has no parent, ie, it was not imported. When it does get imported, main.start() must be invoked by the importer - exactly as we need to happen in the case of integration tests. This allows each integration test to run in isolation which brings lots of benefits.

The fact that we get code coverage for integration tests is a bonus.

Reply:
Not sure if this is what Kenny means, but I just thought we could avoid the `if(!module.parent)` if we split this into two files: one that just exports the application object, another that runs like a script, importing the application and running it.

The `if (!module.parent)` is pretty standard but better avoided IMO.

Reply:
Yup, that's what I meant. Probably just move all the server composition code to `src/server.ts` or something. Then you can leave src/index.ts as the main entry point to start the app.

Reply:
Awesome! Then let's do that. Since Warren's already using `app`, I think we could go with that for the object, and keep `index.ts` as the script that imports and runs it.